[luileng] iP

luileng opened the pull request:
# Duck 🦆

> “Your mind is for having ideas, not holding them.” – David Allen

## Summary

Duck frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ **SUPER FAST** to use

All you need to do is,
1. Download it from [here](https://github.com/luileng/ip).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉

## Progress
- [X] Week 2
- [X] Week 3
- [x] Week 4

## Code Snippet

From `ip/src/main/java/duck/main/Duck.java`
```java
public class Duck {
    public static void main(String[] args) {
        new Duke("data/duck.txt").run();
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

luileng committed:
Add Duck.java

luileng committed:
User command echoed

luileng committed:
Add list

luileng committed:
Add mark and unmark

luileng committed:
Add different types of tasks

luileng committed:
Add test

luileng committed:
Add error handling

luileng committed:
Add delete

luileng committed:
Add javadoc comments

luileng committed:
Add save and load tasks

luileng committed:
Merge branch-Level-7

luileng committed:
Add date to deadline

luileng committed:
Merge branch-Level-8

luileng committed:
Add more OOP

luileng committed:
Merge branch-A-MoreOOP

luileng committed:
Add packages

luileng committed:
Merge branch-A-Packages

luileng committed:
Merge branch 'add-gradle-support'

tayxuenye commented on src/main/java/duck/command/DeleteCommand.java:
```diff
@@ -0,0 +1,21 @@
+import duck.storage.Storage;
+
+public class DeleteCommand implements Command {
+    private final String fullCommand;
```
Good intuitive naming

tayxuenye commented on src/main/java/duck/command/ListCommand.java:
```diff
@@ -0,0 +1,19 @@
+    }
+
+    @Override
+    public void execute(TaskList list, Ui ui, Storage storage) {
```
A bit ambiguous naming to me. Hard to figure what "execute" does.

tayxuenye commented on src/main/java/duck/command/TaskCommand.java:
```diff
@@ -0,0 +1,57 @@
+            return new ToDo(fullCommand.split(" ", 2)[1]);
+        } catch (EmptyToDoException e) {
+            System.out.println(e.getMessage());
+            return null;
```
Will it cause a null pointer exception?

tayxuenye commented on src/main/java/duck/ui/Ui.java:
```diff
@@ -0,0 +1,71 @@
+        return scanner.nextLine();
+    }
+
+    public void showWelcomeMessage() {
```
Good intuitive naming!

tayxuenye reviewed (commented):
Good job!

luileng committed:
Add gradle

luileng committed:
Merge branch 'branch-A-Gradle'

luileng committed:
Add junit tests

luileng committed:
Merge branch-A-JUnit

luileng committed:
Add jar file

luileng committed:
Merge branch 'branch-A-Jar'

luileng committed:
Add javadocs

luileng committed:
Ensure coding standard

luileng committed:
Add find command

luileng committed:
Merge branch 'branch-A-JavaDoc'

luileng committed:
Merge branch 'branch-A-CodingStandard'

luileng committed:
Merge branch-Level-9

limyixiang commented on src/main/java/duck/storage/Storage.java:
```diff
@@ -0,0 +1,77 @@
+package duck.storage;
+
+import duck.command.InvalidCommandException;
+import duck.task.*;
```
Please explicitly import all of the classes that you use here!

limyixiang commented on src/main/java/duck/storage/Storage.java:
```diff
@@ -0,0 +1,77 @@
+                String line = fileScanner.nextLine();
+                String[] parts = line.split(" \\| ");
+                Task task = switch (parts[0]) {
+                    case "T" -> new ToDo(parts[2]);
```
I think case has the wrong level of indentation here: 

> Note there is no indentation for case clauses.

limyixiang commented on src/main/java/duck/task/EmptyToDoException.java:
```diff
@@ -0,0 +1,7 @@
+package duck.task;
+
+public class EmptyToDoException extends Exception {
```
Please add a JavaDoc statement for your class!

limyixiang commented on src/test/java/duck/task/DeadlineTest.java:
```diff
@@ -0,0 +1,19 @@
+
+public class DeadlineTest {
+    @Test
+    public void testStringConversion(){
```
Please add a whitespace " " before your `{`!
`public void testStringConversion() {`

limyixiang commented on src/test/java/duck/task/EventTest.java:
```diff
@@ -0,0 +1,13 @@
+
+public class EventTest {
+    @Test
+    public void testFileStringConversion(){
```
Please add a whitespace " " before your `{`!
`public void testFileStringConversion() {`

limyixiang reviewed (commented):
Great work!

luileng committed:
Add checkstyle config

luileng committed:
Merge branch-A-Checkstyle

starbucksventi commented on src/main/java/duck/command/DeleteCommand.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    @Override
+    public void execute(TaskList list, Ui ui, Storage storage) {
```
Looks good to merge! But you might need to add JavaDoc for this method.

starbucksventi commented on src/main/java/duck/command/TaskCommand.java:
```diff
@@ -0,0 +1,91 @@
+     */
+    @Override
+    public void execute(TaskList list, Ui ui, Storage storage) {
+        Task task = switch (commandType) {
```
Elegant use of switch statement! Small formatting change though - the case likely should not be indented based on this: https://se-education.org/guides/conventions/java/intermediate.html

starbucksventi commented on src/main/java/duck/parser/Parser.java:
```diff
@@ -0,0 +1,40 @@
+    public static Command parse(String fullCommand) throws InvalidCommandException {
+        String commandType = Parser.getCommandType(fullCommand);
+        return switch (commandType) {
+            case "todo", "deadline", "event" -> new TaskCommand(fullCommand, commandType);
```
Compact and elegant style! However, here too the case likely should not be indented, as shown here: https://se-education.org/guides/conventions/java/intermediate.html

luileng committed:
Add gui for duck echo

luileng committed:
Merge branch-Level-10

luileng committed:
Add functionality to gui

luileng committed:
Merge branch-Add-Functionality-To-GUI

starbucksventi reviewed (commented):
Apologies, forgot to leave an overall comment earlier, so here it is: Kudos on the extensive use of OOP and packages to delegate away specialised functionalities. Other than a few nits, looks good to merge!

luileng committed:
Add assertions

The current code does not have assertions.

Assertions can help me to identify mistakes in the code.

Implement assertions to define assumptions about the program state
such as preconditions, postconditions and class invariants.

This is so that the runtime can verify them and assertions act
complimentarily with Exceptions.

luileng committed:
Improve code quality

Some parts of the code may be difficult to read due to its structure.

This affects its clarity especially for other programmers.

Add line spacing in code so that it adheres to the logical structure.

Doing so makes it more readable and easier to follow.

luileng committed:
Merge pull request #1 from luileng/branch-A-Assertions

Merge branch-A-Assertions

luileng committed:
Resolve merge conflicts with master

luileng committed:
Merge pull request #2 from luileng/branch-A-CodeQuality

Improve code quality

luileng committed:
Set up CI

The project lacks automated CI processes.

This increases the risk of bugs and integration issues. It also
requires more manual testing.

Add a GitHub Actions workflow for Continuous Integration to automate
build and test processes.

This setup ensures automated testing and integration, improving code
quality and reducing manual testing efforts.

luileng committed:
Merge branch-A-CI

luileng committed:
Add remind command

luileng committed:
Merge branch 'branch-BCD-extension'

luileng committed:
Add product website

luileng committed:
Resolve checkstyle errors

luileng committed:
Merge branch 'branch-Resolve-checkstyle-errors'

luileng committed:
Update test for text-ui-test

luileng committed:
Update test to mark

luileng committed:
Merge branch 'branch-Update-Test'

luileng committed:
Update gradle

luileng committed:
Merge branch 'branch-Update-Gradle'

luileng committed:
Add user guide

luileng committed:
Merge branch 'branch-A-UserGuide'

luileng committed:
Improve header comments

luileng committed:
Merge branch 'branch-Improve-Header-Comments'

luileng committed:
Add parser test

luileng committed:
Merge branch 'branch-Parser-Test'

luileng committed:
Add invalid date format exception

luileng committed:
Merge branch 'branch-A-MoreErrorHandling'

luileng committed:
Improve abstraction

luileng committed:
Merge branch 'branch-Improve-Abstraction'

luileng committed:
Add welcome message

luileng committed:
Merge branch 'branch-Add-Welcome-Message'

luileng committed:
Add error handling for tasks

luileng committed:
Improve code style

luileng committed:
Merge branch 'branch-Improve-Code-Style'
